### PR TITLE
RavenDB-8088 - licensing in a cluster

### PR DESCRIPTION
**Pull request by grisha-kotler**
http://issues.hibernatingrhinos.com/issue/RavenDB-8088

### CODE REVIEW

File: src/Raven.Studio/typescript/commands/licensing/deactivateLicenseCommand.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import commandBase = require("commands/commandBase");
+import endpoints = require("endpoints");
+
+class deactivateLicenseCommand extends commandBase {
+
+    constructor() {
```
Comment (ml054) [inline src/Raven.Studio/typescript/commands/licensing/deactivateLicenseCommand.ts]: no need for this constructor :)

File: src/Raven.Studio/typescript/viewmodels/shell/about.ts [hunk 2]
```diff
@@ -2,12 +2,14 @@ import viewModelBase = require("viewmodels/viewModelBase");
...
 import license = require("models/auth/license");
 import registration = require("viewmodels/shell/registration");
+import deactivateLicenseCommand = require("commands/licensing/deactivateLicenseCommand");
 
 class about extends viewModelBase {
 
     clientVersion = shell.clientVersion;
     serverVersion = shell.serverBuildVersion;
     licenseStatus = license.licenseStatus;
+    deactivating = ko.observable<boolean>(false);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/shell/about.ts]: by convention in v4.0 we use variable spinners for holding spinners see:
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Studio/typescript/viewmodels/database/indexes/indexes.ts#L38

Comment (grisha-kotler) [reply src/Raven.Studio/typescript/viewmodels/shell/about.ts]: done

File: src/Raven.Studio/typescript/commands/licensing/deactivateLicenseCommand.ts (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Studio/typescript/commands/licensing/deactivateLicenseCommand.ts]: done

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 3]
```diff
@@ -140,17 +174,39 @@ public LicenseStatus GetLicenseStatus()
             }
         }
 
-        public void Activate(License license, bool skipLeaseLicense)
+        public async Task Activate(License license, bool skipLeaseLicense)
         {
+            if (_serverStore.CurrentState == RachisConsensus.State.Passive)
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseManager.cs:179]: Not a good idea, one of the first things that an admin will want to do is to set the license, you want this to call `EnsureNotPasssive`

File: src/Raven.Server/Commercial/LicenseValidator.cs [hunk 4]
```diff
@@ -39,7 +39,7 @@ private static byte[] GetBytesFromBase64String(string str)
             return Convert.FromBase64String(str);
         }
 
-        public static Dictionary<string, object> Validate(License licenseKey, RSAParameters rsAParameters)
+        public static Dictionary<string, object> Validate(License licenseKey, RSAParameters rsaParameters)
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseValidator.cs:42]: Shouldn't be using `sodium` here to do this? That would be simpler to handle, I think.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 5]
```diff
@@ -100,6 +101,7 @@ protected override void Apply(TransactionOperationContext context, BlittableJson
                         break;
 
                     case nameof(DeleteValueCommand):
+                    case nameof(DeactivateLicenseCommand):
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:104]: Shouldn't we also have some code to handle the behavior of license disactivation?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 6]
```diff
@@ -171,6 +173,9 @@ protected override void Apply(TransactionOperationContext context, BlittableJson
                     case nameof(ToggleSubscriptionStateCommand):
                         SetValueForTypedDatabaseCommand(context, type, cmd, index, leader);
                         break;
+                    case nameof(PutLicenseCommand):
+                        PutValue<License>(context, type, cmd, index, leader);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:177]: Where is the logic for handling licenses?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 7]
```diff
@@ -389,12 +394,14 @@ private void DeleteValue(TransactionOperationContext context, string type, Blitt
                 if (delCmd.Name.StartsWith("db/"))
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:394]: What isn't this using a constant? And it probably need to use case insensitive match.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 8]
```diff
@@ -412,7 +419,7 @@ private void PutValue<T>(TransactionOperationContext context, string type, Blitt
                 if (command.Name.StartsWith(Constants.Documents.Prefix))
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:419]: Probably need case insensitive match

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 9]
```diff
@@ -528,6 +535,10 @@ private void OnValueChanged(object sender, (long Index, string Type) t)
                 case nameof(PutClientConfigurationCommand):
                     _lastClientConfigurationIndex = t.Index;
                     break;
+                case nameof(PutLicenseCommand):
+                case nameof(DeactivateLicenseCommand):
+                    LastLicenseIndex = t.Index;
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs:540]: What is going to read this ?

File: src/Raven.Server/Web/Studio/LicenseHandler.cs [hunk 10]
```diff
@@ -48,9 +49,17 @@ public Task Activate()
                 license = JsonDeserializationServer.License(json);
```
Comment (ayende) [inline src/Raven.Server/Web/Studio/LicenseHandler.cs:49]: Need to handle validation here.

**Pull request merged**